File: pset6/credit/credit.py
def main():
    while True:
        nbr = input("cartao: ")

        if nbr.isdigit() and int(nbr) > 0:
            nbr = int(nbr)
            break

    luhn = cs_luhn(nbr)
    operadora = cs_operadoras(nbr)

    if luhn == 0 or operadora == 0:
        print("INVALID")
    elif operadora == 1:
        print("AMEX")
    elif operadora == 2:
        print("MASTERCARD")
    elif operadora == 3:
        print("VISA")


def cs_digits(nbr):
    qty = 0

    while nbr > 0:
        qty += 1
        nbr //= 10

    return qty


def cs_luhn(nbr):
    multiplos = 0
    simples = 0

    while nbr > 0:
        simples += nbr % 10
        nbr = nbr // 10
        digit = (nbr % 10) * 2
        if digit > 9:
            digit = (digit % 10) + ((digit // 10) % 10)
        multiplos += digit
        nbr = nbr // 10

    total = multiplos + simples

    if total % 10 == 0:
        return True
    else:
        return False


def cs_operadoras(nbr):
    len = cs_digits(nbr)

    while nbr >= 100:
        nbr //= 10

    if (nbr == 34 or nbr == 37) and len == 15:
        return 1
    elif (nbr >= 51 and nbr <= 55) and len == 16:
        return 2
    elif (nbr >= 40 and nbr <= 49) and (len == 13 or len == 16):
        return 3
    else:
        return 0


main()

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int ms_digits(long nbr);
int ms_luhn(long nbr);
int ms_operadoras(long nbr);

int main(void)
{
	long nbr;
	int luhn;
	int operadora;

	do
	{
		nbr = get_long("cartao: ");
	}
	while (nbr < 1);
	luhn = ms_luhn(nbr);
	operadora = ms_operadoras(nbr);
	if (luhn == 0 || operadora == 0)
		printf("INVALID\n");
	else if (operadora == 1)
		printf("AMEX\n");
	else if (operadora == 2)
		printf("MASTERCARD\n");
	else if (operadora == 3)
		printf("VISA\n");
	return (0);
}

int ms_digits(long nbr)
{
	int qty;

	qty = 0;
	while (nbr > 0)
	{
		qty++;
		nbr /= 10;
	}
	return (qty);
}

int ms_luhn(long nbr)
{
	int multiplos;
	int simples;
	int digit;
	int total;

	multiplos = 0;
	simples = 0;
	while (nbr > 0)
	{
		simples += nbr % 10;
		nbr = nbr / 10;
		digit = (nbr % 10) * 2;
		if (digit > 9)
			digit = (digit % 10) + ((digit / 10) % 10);
		multiplos += digit;
		nbr = nbr / 10;
	}
	total = multiplos + simples;
	if (total % 10 == 0)
		return (1);
	else
		return (0);
}

int ms_operadoras(long nbr)
{
	int len;

	len = ms_digits(nbr);
	while (nbr >= 100)
		nbr /= 10;
	if ((nbr == 34 || nbr == 37) && len == 15)
		return (1);
	else if ((nbr >= 51 && nbr <= 55) && len == 16)
		return (2);
	else if ((nbr >= 40 && nbr <= 49) && (len == 13 || len == 16))
		return (3);
	else
		return (0);
}
